### PR TITLE
:sparkles: feat/add-heptagram-api

### LINKED ISSUE #1234: [FEAT] - Add Heptagram Scamlinks api for scam detection and reporting
### What feature would you like to see?

Use the [Heptagram API](https://github.com/Heptagram-Bot-Project/api)'s scamlink endpoints in addition to walshy's endpoint to check for & report spam links.

I develop this API, so am happy to help, and/or PR this.

### Additional information

Conversed with Naomi about this in discord.

contributing - Discord Discord 2rUrf5S9

[https://discord.com/channels/778130114772598785/869607198450389062/963196223483969556](https://discord.com/channels/778130114772598785/869607198450389062/963196223483969556)

### PR DESCRIPTION
**Pull request by j-dogcoder**
# Pull Request

## Related Issue

Closes #1234 

## Documentation

For _any_ version updates, please verify if the [documentation page](https://docs.beccalyria.com?utm_source=github&utm_medium=pr-template) needs an update. If it does, please [create an issue there](https://github.com/BeccaLyria/discord-documentation/issues/new?assignees=nhcarrigan&labels=%F0%9F%9A%A6+status%3A+awaiting+triage&template=update.md&title=%5BUPDATE%5D) to ensure it is not forgotten.

- [ ] My contribution does NOT require a documentation update.
- [x] My contribution DOES require a documentation update.

### CODE REVIEW

**Comment by j-dogcoder** [comment]
I think this works @nhcarrigan.

File: src/listeners/automod/automodPhish.ts [hunk 1]
```diff
@@ -47,6 +47,23 @@ export const automodPhish: ListenerHandler = async (
     let scamSource = "";
 
     for (const link of blockedLinkList) {
+
+      const checkHeptagramAPI = await axios.get<boolean>(
+        `http://heptagrambotproject.com/api/v0/api/scam/link/check?url=${link}`,
+        {
+          // send authentication header
+          headers: {
+            Authorization: "Bearer " + config.heptagramApiToken,
```
Comment (nhcarrigan) [inline src/listeners/automod/automodPhish.ts:56]: So the `config` object is the server's settings from the database.

Becca's ENV values are in `Becca.configs` - we'll need to validate the API key is present and mount it to this object in the `validateEnv` module.

File: src/modules/naomi/naomiAntiphish.ts [hunk 2]
```diff
@@ -18,7 +18,19 @@ export const naomiAntiphish = async (Becca: BeccaLyria, message: Message) => {
     const result = await axios.post<{ message: string }>(
       "https://bad-domains.walshy.dev/report",
       { domain: link }
-    );
+    ) &&
+      axios.post<boolean>(
+      `http://heptagrambotproject.com/api/v0/api/scam/link/post}`,
```
Comment (nhcarrigan) [inline src/modules/naomi/naomiAntiphish.ts:23]: Is that `}` intended?
```suggestion
      `http://heptagrambotproject.com/api/v0/api/scam/link/post`,
```

**Review by nhcarrigan — CHANGES_REQUESTED** [review]
This is a good start. Left a couple of notes below.

I'll need to re-evaluate the flow for reporting links that get through - I'll handle that later on, but leaving a note here for myself to do so.

File: src/listeners/automod/automodPhish.ts (hunk 1 above)
Comment (nhcarrigan) [reply src/listeners/automod/automodPhish.ts:56]: ENV Validation: https://github.com/BeccaLyria/discord-bot/blob/main/src/modules/validateEnv.ts

Type-def: https://github.com/BeccaLyria/discord-bot/blob/main/src/interfaces/BeccaLyria.ts

File: src/modules/naomi/naomiAntiphish.ts (hunk 2 above)
Comment (j-dogcoder) [reply src/modules/naomi/naomiAntiphish.ts:23]: Woops!

File: src/listeners/automod/automodPhish.ts (hunk 1 above)
Comment (j-dogcoder) [reply src/listeners/automod/automodPhish.ts:56]: Should be fixed now.

**Comment by j-dogcoder** [comment]
Ready for review again @nhcarrigan!

File: src/listeners/automod/automodPhish.ts [hunk 3]
```diff
@@ -47,6 +47,23 @@ export const automodPhish: ListenerHandler = async (
...
+          },
+        });
```
Comment (nhcarrigan) [inline src/listeners/automod/automodPhish.ts:58]:
```suggestion
    for (const link of blockedLinkList) {
      const checkHeptagramAPI = await axios.get<boolean>(
        `http://heptagrambotproject.com/api/v0/api/scam/link/check?url=${link}`,
        {
          // send authentication header
          headers: {
            Authorization: "Bearer " + Becca.configs.heptagramApiToken,
          },
        }
      );
```
Linter fix + update the API token location.

File: src/modules/naomi/naomiAntiphish.ts [hunk 4]
```diff
@@ -18,7 +18,19 @@ export const naomiAntiphish = async (Becca: BeccaLyria, message: Message) => {
...
+      {
+      headers: {
+        Authorization: "Bearer " + config.heptagramApiToken,
+      },
+      body: {
+        link: link,
+        reportedby: `${message.author.id} || Becca Lyria`,
+      },
+    });
+    ;
```
Comment (nhcarrigan) [inline src/modules/naomi/naomiAntiphish.ts:33]:
```suggestion
    const heptagramResult = await axios.post<boolean>(
      `http://heptagrambotproject.com/api/v0/api/scam/link/post`,
      {
        headers: {
          Authorization: "Bearer " + Becca.configs.heptagramApiToken,
        },
        body: {
          link: link,
          reportedby: `${message.author.id} || Becca Lyria`,
        },
      }
    );
```
Separate these into two calls.

File: src/modules/naomi/naomiAntiphish.ts [hunk 5]
```diff
@@ -18,7 +18,19 @@ export const naomiAntiphish = async (Becca: BeccaLyria, message: Message) => {
     const result = await axios.post<{ message: string }>(
       "https://bad-domains.walshy.dev/report",
       { domain: link }
```
Comment (nhcarrigan) [inline src/modules/naomi/naomiAntiphish.ts:20]:
```suggestion
    const walshyResult = await axios.post<{ message: string }>(
      "https://bad-domains.walshy.dev/report",
      { domain: link }
    );

```
Separate these into two calls.

File: src/modules/naomi/naomiAntiphish.ts [hunk 6]
```diff
@@ -18,7 +18,19 @@ export const naomiAntiphish = async (Becca: BeccaLyria, message: Message) => {
...
+      },
+      body: {
+        link: link,
+        reportedby: `${message.author.id} || Becca Lyria`,
+      },
+    });
+    ;
 
     await message.reply(
       `I have reported that domain! Here's the result:\n\`${result.data.message}\``
```
Comment (nhcarrigan) [inline src/modules/naomi/naomiAntiphish.ts:36]:
```suggestion
      `I have reported that domain! Here's the result:\n${JSON.stringify({
        walshy: walshyResult.data.message,
        heptagram: heptagramResult.data,
      })}`
```
And update the response structure to match.

**Review by nhcarrigan — CHANGES_REQUESTED** [review]
These should fix the linter stuff.

File: src/modules/validateEnv.ts [hunk 7]
```diff
@@ -71,6 +71,10 @@ export const validateEnv = (
       return { valid: false, message: "Missing Orbit API key" };
     }
 
+    if (!process.env.HEPTAGRAM_API_TOKEN) {
+      return { valid: false, message: "Missing Heptagram API token"};
```
Comment (nhcarrigan) [inline src/modules/validateEnv.ts:75]:
```suggestion
      return { valid: false, message: "Missing Heptagram API token" };
```

File: src/modules/validateEnv.ts [hunk 8]
```diff
@@ -93,6 +97,7 @@ export const validateEnv = (
       voteChannel: process.env.VOTE_CHANNEL_ID,
       habiticaKey: process.env.HABITICA_KEY,
       orbitKey: process.env.ORBIT_KEY,
+      heptagramApiToken: process.env.HEPTAGRAM_API_TOKEN
```
Comment (nhcarrigan) [inline src/modules/validateEnv.ts:100]:
```suggestion
      heptagramApiToken: process.env.HEPTAGRAM_API_TOKEN,
```

**Review by nhcarrigan — CHANGES_REQUESTED** [review]
I'll go ahead and commit these so we can get this in.

**Review by nhcarrigan — APPROVED** [review]
Alright, we're good to go!

**Comment by nhcarrigan** [comment]
## Congratulations

Congratulations on making your first pull request to our project. We have reviewed your contributions and are happy to accept them.

We hope you enjoyed the contribution process, and encourage you to watch for issues labelled `help wanted` for additional opportunities to contribute.

**Comment by j-dogcoder** [comment]
Yay! Is the report logic working now? Or do you have to fix that still?

**Pull request merged**